pytorch_lightning/plugins/training_type/training_type_plugin.py

```diff
@@ -153,7 +153,10 @@ def results(self) -> Optional[Union[_EVALUATE_OUTPUT, _PREDICT_OUTPUT]]:
 
     def load_checkpoint(self, checkpoint_path: _PATH) -> Dict[str, Any]:
         torch.cuda.empty_cache()
-        return self.checkpoint_io.load_checkpoint(checkpoint_path)
+        checkpoint = self.checkpoint_io.load_checkpoint(checkpoint_path)
+        self.lightning_module.on_load_checkpoint(checkpoint)
+        self.load_model_state_dict(checkpoint)
```

Comment:
n00b question: do you think this ought to be on the base class? vs calling `self.lightning_module.on_load_checkpoint(checkpoint)` in `load_model_state_dict` ?

Reply:
do you mean if `on_load_checkpoint` should be in `load_model_state_dict` v.s. `load_checkpoint` here in ttp base class? if so I was thinking to maybe keep `load_model_state_dict` only performing `load_state_dict` as is, wdyt?

Reply:
`load_checkpoint` is materializing the file contents in memory whereas `load_model_state_dict` is specifically on reloading the model's state

in my head, i think `on_load_checkpoint` is in spirit closer to `load_model_state_dict` than it is to loading the file...but i'm not 100% certain if my assumption is right. that's based on the current codebase where `model.on_load_checkpoint` is called right before the state dict.

if this was redone, maybe the name for this hook would've been `on_load_state_dict` ..

@tchaton @awaelchli does my understanding sqaure up with yours?

Reply:
hm agree `on_load_checkpoint` is closer to `load_model_state_dict` than `load_checkpoint (file)` - maybe some renaming?
- old `load_checkpoint` --> `load_checkpoint_file`
- new method `load_checkpoint_file_and_model_state` (a better name than this?) = `load_checkpoint_file`, `on_load_checkpoint`, `load_model_state_dict`

Reply:
just to confirm - like the below? (so `load_model_state_dict`, `load_checkpoint` will have different functionalities across different TTPs)

base TTP:
- `load_model_state_dict` = `on_load_checkpoint`, `model.load_state_dict`
- `load_checkpoint` = load checkpoint file

serialized loading TTPs:
- `load_model_state_dict` = `model.load_state_dict` only
- `load_checkpoint` = for rank in ranks, `super().load_checkpoint`, `on_load_checkpoint`, `load_model_state_dict`
- (`load_model_state_dict` will be called again in checkpoint connector, but early exits with the `_self_deleted_checkpoint_state_dict` boolean)

Reply:
yes that's correct, at least this is my preference. what do you think @jjenniferdai @awaelchli @carmocca ?

Reply:
I slightly prefer the second option because it's more consistent across TTPs. but defer to others more experienced :)

Reply:
was pinged here but sorry, I'm not able to follow the conversation.
- you can rename things in the io plugin because its not released yet
- we are not allowed to change the order of hook calls

Reply:
no change to order of hook calls or io plugin, but moving calling responsibility of `LightningModule.on_load_checkpoint` from checkpoint connector to TTP ("to work through" portion of issue #9406)

2 different options in ([comment](https://github.com/PyTorchLightning/pytorch-lightning/pull/9605#discussion_r717806551)) on how to organize this:
- first option = this [comment](https://github.com/PyTorchLightning/pytorch-lightning/pull/9605#discussion_r717165512)
- second option = this PR currently